caddytls/ctmonitor.go

```diff
@@ -0,0 +1,298 @@
+)
+
+var (
+	filePath     string = filepath.Join(caddy.AssetsPath(), "ct_id")
+	ctConfigFile string = filepath.Join(caddy.AssetsPath(), "ct_config")
+)
+
+// CtConfig struct will allow me to get the config data from a file
+type CtConfig struct {
+	IncludeSubdomains bool `json:"subdomains"`
+	IncludeWildCards  bool `json:"wildCards"`
+}
+
+// CertSpotterResponse is the json response formatting that is used in the program.
+type CertSpotterResponse struct {
+	ID           string   `json:"id"`
+	TBSSHA256    string   `json:"tbs_sha256"`
+	DNSNames     []string `json:"dns_names"`
+	PubKeySha256 string   `json:"pubkey_sha256"`
+	Issuer       struct {
+		Name         string `json:"name"`
+		PubKeySha256 string `json:"pubkey_sha256"`
+	} `json:"issuer"`
+	NotBefore string `json:"not_before"`
+	NotAfter  string `json:"not_after"`
+	Cert      struct {
+		ID     string `json:"id"`
+		Type   string `json:"type"`
+		SHA256 string `json:"sha256"`
+		Data   string `json:"data"`
+	} `json:"cert"`
+}
+
+// QueryConfig is the configuration information for my queries.
+type QueryConfig struct {
+	Subdomains bool
+	WildCards  bool
+	Query      string
+	Index      int
+}
+
+// CompareCerts compares the certificates that caddy is serving against the certificates
+// that certSpotter has found, if there are any that don't match the caddy certificates,
+// they are reported to the user.
+func CompareCerts(caddyCerts map[string]struct{}, certSpotterCerts map[string]string) {
+	for key := range certSpotterCerts {
+		if _, ok := caddyCerts[key]; !ok {
+			log.Printf("[WARNING] Certificate found that caddy is not monitoring, issued by: %v\n", certSpotterCerts[key])
+		}
+	}
+}
+
+// getCaddyCerts retrieves the certificates that caddy monitors and returns them
+// as a map with the key being the bytes of the certificate cast to a string.
+func getCaddyCerts() ([]string, map[string]struct{}) {
+	var (
+		caddyCerts = make(map[string]struct{}) //caddyCerts consists of the certificates that Caddy is serving.
+		caddyDNS   = make([]string, 0, 10)     //caddyDNS consists of the Subject Alternate Names that Caddy is hosting.
+	)
+	for _, inst := range caddy.Instances() {
+		inst.StorageMu.RLock()
+		certCache, ok := inst.Storage[CertCacheInstStorageKey].(*certificateCache)
+		inst.StorageMu.RUnlock()
+		if !ok || certCache == nil {
+			continue
+		}
+		certCache.RLock()
+		for _, certificate := range certCache.cache {
+			caddyDNS = append(caddyDNS, certificate.Names...)
+			certBytes := string(certificate.Certificate.Certificate[0])
+			if _, ok := caddyCerts[certBytes]; !ok {
+				caddyCerts[certBytes] = struct{}{}
+			}
+		}
+		certCache.RUnlock()
+	}
+	return caddyDNS, caddyCerts
+}
+
+func getLatestIndex(fileName string) (int, error) {
+	fmt.Printf("ct_id FilePath: %v\n", fileName)
+	indexBytes, err := ioutil.ReadFile(fileName)
+	if os.IsNotExist(err) {
+		log.Println("getLatestIndex failed, could not find the file")
```

Comment:
Remember the error handling pattern: `return fmt.Errorf("could not load last index file: %v", err)`